app/src/main/java/com/javahelp/frontend/activity/ProviderRegistrationActivity.java

```diff
@@ -25,14 +50,109 @@ protected void onCreate(Bundle savedInstanceState) {
+        providerRegistrationViewmodel.setPassword1(binding.password.getText().toString());
+        providerRegistrationViewmodel.setPassword2(binding.repassword.getText().toString());
+        providerRegistrationViewmodel.setProviderUserInfo(binding.email.getText().toString(),
+                binding.firstname.getText().toString(),
+                binding.lastname.getText().toString(),
+                binding.home.getText().toString(),
+                binding.phonenumber.getText().toString());
+        registerAttempt();
+    }
+
+    /**
+     * Initiates a register request
+     */
+    private void registerAttempt() {
+        if (binding.username.getText().toString().isEmpty() || binding.password.getText().toString().isEmpty()
+        || binding.repassword.getText().toString().isEmpty()) {
+            providerRegistrationViewmodel.setRegisterError("Please enter username and password");
+            return; // early return on invalid username or password
+        }else if(!binding.password.getText().toString().equals(binding.repassword.getText().toString())){
+            providerRegistrationViewmodel.setRegisterError("Passwords do not match");
+            return;
+            }else if(binding.email.getText().toString().isEmpty() || binding.firstname.getText().toString().isEmpty()
+        || binding.lastname.getText().toString().isEmpty() || binding.home.getText().toString().isEmpty()
+        ||binding.phonenumber.getText().toString().isEmpty()){
+            providerRegistrationViewmodel.setRegisterError("Please enter your personal information");
+            return;
+        }
+    }
+
+    /**
+     * Updates this {@link RegisterResult} to reflect the passed register result
+     *
+     * @param registerResult {@link Optional} {@link RegisterResult} to update on
+     */
+    private void updateOnRegisterResult(Optional<RegisterResult> registerResult) {
+        if(registerResult.isPresent()){
+            updateOnPresentRegisterResult(registerResult.get());
+        } else {
+            binding.registerErrorText.setVisibility(View.GONE);
+        }
+    }
+
+    /**
+     * Updates the UI based on a {@link RegisterResult}. This is called with non-null register results, meaning
+     * a registration has been attempted, and has either succeeded or failed. This is unlike updateOnRegisterResult
+     * which is called with an {@link Optional} {@link RegisterResult}, and calls this method if that
+     * {@link Optional} is present.
+     *
+     * @param result {@link RegisterResult} of login attempt
+     */
+    private void updateOnPresentRegisterResult(RegisterResult result) {
+        if (result.isSuccess()) {
+            binding.registerErrorText.setText("Register successful");
+            storeCredentials(result.getUser(), result.getToken());
+            startActivity(new Intent(ProviderRegistrationActivity.this, FrontPageActivity.class));
```

Comment:
I am not sure what is done on the backend but this means after they register, they are automatically logged in. I don't know whether there will be any token issue associated with this or not. I am thinking that after they register, they can be directed to the login page again and they will have to enter their username and password again. In this case we don't need to worry about token stuff in the register use case.

Reply:
Yes, but what I have discussed with Jacob is that we could automatically log in after the registration, I am not too sure about it because I have to also wait for Jacob to implement the gateway

Reply:
The gateway will send a token back too so there will not be an issue